onnx/defs/nn/defs.cc

```diff
@@ -1534,41 +1555,100 @@ ONNX_OPERATOR_SET_SCHEMA(
             "var",
             "running (training) or estimated (testing) variance tensor of shape (C).",
             "T")
+        .Input(
+            5,
+            "training_mode",
+            "If set to true, run spatial batch normalization in training mode, default is false.",
+            "T1",
+            OpSchema::Optional)
         .Output(0, "Y", "The output tensor of the same shape as X", "T")
         .Output(
             1,
             "mean",
-            "The running mean after the BatchNormalization operator.",
+            "The running mean when training_mode=True, "
+            "or the estimated mean when training_mode=False (Tensor of shape (C))."
+            "Note that this output cannot be an input of any other operator.",
             "T",
             OpSchema::Optional)
         .Output(
             2,
             "var",
-            "The running variance after the BatchNormalization operator.",
+            "The running variance when training_mode=True, "
+            "or the estimated variance when training_mode=False (Tensor of shape (C))."
+            "Note that this output cannot be an input of any other operator.",
             "T",
             OpSchema::Optional)
         .Output(
             3,
             "saved_mean",
-            "Saved mean used during training to speed up gradient "
-            "computation.",
+            "Saved mean used during training to speed up gradient computation (Tensor of shape (C)).",
             "T",
             OpSchema::Optional)
         .Output(
             4,
             "saved_var",
-            "Saved variance used during training to speed up "
-            "gradient computation.",
+            "Saved variance used during training to speed up gradient computation (Tensor of shape (C)).",
             "T",
             OpSchema::Optional)
         .TypeConstraint(
             "T",
             {"tensor(float16)", "tensor(float)", "tensor(double)"},
             "Constrain input and output types to float tensors.")
+        .TypeConstraint(
+            "T1",
+            {"tensor(bool)"},
+            "Constrain input 'training_mode' types to boolean tensors.")
         .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
           propagateShapeAndTypeFromFirstInput(ctx);
-          // TODO in training mode, it may be possible to infer some of
-          // the other outputs as well.
+
```

Comment:
Before we use the i-th inputShape, we should check if an i-th inputShape is available (using either https://github.com/onnx/onnx/blob/121114107c38ec49bab4955c13cb82b7675bfc47/onnx/defs/shape_inference.h#L269 or https://github.com/onnx/onnx/blob/121114107c38ec49bab4955c13cb82b7675bfc47/onnx/defs/shape_inference.h#L274 ). It is possible to have input N, but have no shape for it (if shape is unknown).

Reply:
I updated the shape inference to include the check, thanks!